From b3588cbf8da93145aa053b51cdaf7fcabc984268 Mon Sep 17 00:00:00 2001
From: Jouni Malinen <j@w1.fi>
Date: Wed, 18 Sep 2024 11:12:05 +0530
Subject: [PATCH] UPSTREAM: hostapd: Clear peer_rejected_groups when no element
 is included When parsing a SAE Commit message, the temporary
 peer_rejected_groups parameter was left to its old value in cases where the
 new SAE Commit message did not include the Rejected Groups element. This
 could result in unexpected behavior if a previously processed SAE Commit
 message included a Rejected Groups element that claimed one of the enabled
 groups to be rejected.

Explicitly clear the peer_rejected_groups value when parsing an SAE
Commit message without a Rejected Groups element to avoid rejecting the
new message based on some previously received incorrect information.
This avoids some potential denial-of-service issues during the lifetime
of the SAE temporary data.

Tag: SAE
Patch-work: None
Patch-dependency: None

Signed-off-by: Jouni Malinen <j@w1.fi>
Signed-off-by: Arunpandi Kannan <quic_arunpand@quicinc.com>
---
 src/common/sae.c | 8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git a/src/common/sae.c b/src/common/sae.c
index a9acb7d..e1ed323 100644
--- a/src/common/sae.c
+++ b/src/common/sae.c
@@ -2110,8 +2110,11 @@ static int sae_parse_rejected_groups(struct sae_data *sae,
 
 	wpa_hexdump(MSG_DEBUG, "SAE: Possible elements at the end of the frame",
 		    *pos, end - *pos);
-	if (!sae_is_rejected_groups_elem(*pos, end))
+	if (!sae_is_rejected_groups_elem(*pos, end)) {
+		wpabuf_free(sae->tmp->peer_rejected_groups);
+		sae->tmp->peer_rejected_groups = NULL;
 		return WLAN_STATUS_SUCCESS;
+	}
 
 	epos = *pos;
 	epos++; /* skip IE type */
@@ -2203,6 +2206,9 @@ u16 sae_parse_commit(struct sae_data *sae, const u8 *data, size_t len,
 		res = sae_parse_rejected_groups(sae, &pos, end);
 		if (res != WLAN_STATUS_SUCCESS)
 			return res;
+	} else {
+		wpabuf_free(sae->tmp->peer_rejected_groups);
+		sae->tmp->peer_rejected_groups = NULL;
 	}
 
 	/* Optional Anti-Clogging Token Container element */
-- 
2.34.1

